devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,191 @@
+import sympy
+import numpy as np
+
+from devito.finite_differences import generate_indices
+
+__all__ = ['Coefficients', 'Coefficient', 'default_rules']
+
+
+# FIXME: Class naming a bit unclear
+class Coefficients(object):
+    """
+    Devito class for users to define custom finite difference weights.
+    Input must be given as Devito Coefficient objects.
+
+    Coefficients objects created in this manner must then be
+    passed to Devito equation objects for the replacement rules
+    to take effect.
+    """
+
+    def __init__(self, *args):
+
+        if any(not arg.is_Coefficient for arg in args):
+            raise TypeError("Non Coefficient object within input")
+
+        self.data = args
+        self.function_list = self.function_list()
+        self.rules = self.rules()
+
+    def function_list(self):
+        function_list = ()
+        for d in self.data:
+            function_list += (d.function,)
+        return list(set(function_list))
+
+    def rules(self):
+
+        def generate_subs(d):
+
+            deriv_order = d.deriv_order
+            function = d.function
+            dim = d.dimension
+            coeffs = d.coefficients
+
+            fd_order = len(coeffs)-1
+
+            side = function.get_side(dim, deriv_order)
+            stagger = function.get_stagger(dim, deriv_order)
+
+            subs = {}
+
+            indices, x0 = generate_indices(function, dim, dim.spacing, fd_order,
+                                           side=side, stagger=stagger)
+
+            for j in range(len(coeffs)):
+                subs.update({function.fd_coeff_symbol()
+                             (indices[j], deriv_order, function, dim): coeffs[j]})
+
+            return subs
+
+        # Figure out when symbolic coefficients can be replaced
+        # with user provided coefficients and, if possible, generate
+        # replacement rules
+        rules = {}
+        for d in self.data:
+            if isinstance(d.coefficients, np.ndarray):
+                rules.update(generate_subs(d))
+
+        return rules
+
+
+class Coefficient(object):
+    """
+    Prepare custom FD coefficients to pass to Coefficients object.
+
+    Parameters
+    ----------
+    deriv_order : represents the order of the derivative being taken.
+    function : represents the function for which the supplied coefficients
+               will be used.
+    dimension : represents the dimension with respect to which the
+                derivative is being taken.
+    coefficients : represents the set of finite difference coefficients
+                  intended to be used in place of the standard
+                  coefficients (obtained from a Taylor expansion).
+    """
+
+    def __init__(self, deriv_order, function, dimension, coefficients):
```

Comment:
perhaps use the same parameters ordering used in the various FD functions, to make it more consistent?